### ISSUE TITLE
Add command line flag for doing a full "regular" sync

### ISSUE THREAD

**Issue by pipermerriam** [ethereum/py-evm#1480]
### What is wrong?

It should be possible to perform a "regular" sync. Meaning sequential download of blocks and sequential application of those blocks to build the chain state.

### How can it be fixed

First, we need to fix an internal naming issue where we refer to our *fast* sync using `SYNC_FULL`.  This should be renamed to `SYNC_FAST`, and use the *fast* terminology in place of *full* everywhere that is currently used.

Then, add a new option for `SYNC_FULL` which just does a regular block-by-block sync.  The cli should support the following:

- `--sync-mode=light/full/fast` with `fast` being the default.
- `--light` as a shortcut for `--sync-mode=light` (already supported)
- `--fast` as a shortcut for `--sync-mode=fast`
- `--full` as a shortcut for `--sync-mode=full`

The `trinity.sync` module will need to be re-organized as well, moving `trinity.sync.full` to `trinity.sync.fast` and adding the new syncer under `trinity.sync.full`.

**Comment by carver** [comment]
Note that the "first" and "then" steps above should probably be separate PRs to make sure all tests pass after the first step.

Hm, geth used to have those shortcuts (eg~ `--fast`) and seems to have dropped them since. We should consider not adding them in the first place.

**Comment by cburgdorf** [comment]
Am I right to assume that we still need to ensure we have the entire state for the current head before we start applying blocks with their transactions (how should it work without that :thinking:?)

**Comment by pipermerriam** [comment]
> how should it work without that

Yes, regular sync should likely do a validate step upon startup (and honestly maybe before applying **any** block to check if the parent block's state root is present in the database so that we can throw a user friendly error in the event that it is not.

**Comment by cburgdorf** [comment]
Cool, I feel like I'm starting to how this called Ethiriuum works.

**Comment by carver** [comment]
It surely already raises an error if the state is missing (_he speculated with out finding it in the code_), but catching it and raising a friendlier message could be reasonable.

**Comment by glaksmono** [comment]
Would like to take a stab on this

**Comment by cburgdorf** [comment]
>but catching it and raising a friendlier message could be reasonable.

I guess we should even go one step further and hint the user that they could recreate the state using something like `--rebuild-state`.

When started with this command, trinity would take the genesis state and then apply all the blocks that it already has until it has build the state up to where its `HEAD` is. Then, from there, the user could run `trinity --full` and get into regular syncing. Does that make sense?

**Comment by cburgdorf** [comment]
According to my simple math, it would take roughly 40 hours to rebuild the entire state for current mainnet (rather than fetch it from other peers)

6751752 / 46 bps* / 60 / 60 = 40.77

* according to our benchmark

But I'm not suggesting this as an alternative to state sync, just as another tool to rebuild the state up to block x (e.g. for testing etc.)

**Comment by carver** [comment]
> I guess we should even go one step further and hint the user that they could recreate the state using something like --rebuild-state.

Sounds interesting. If we have a tool to rebuild the state for them, might as well just run it for them directly instead of telling them to run it.

But I'd say this is all something that could be captured in a separate issue so that this issue can stay a reasonable size.

**Comment by veox** [comment]
> Hm, geth used to have those shortcuts (eg~ `--fast`) and seems to have dropped them since. We should consider not adding them in the first place.

Yes, the shorthands were [deprecated in `geth` v1.6.0](https://github.com/ethereum/go-ethereum/releases/tag/v1.6.0).

Their `--syncmode` was added in commit https://github.com/ethereum/go-ethereum/pull/13875/commits/48339d4ef6dcbcf80f4c74ae64bc7426aa9053c3, part of PR https://github.com/ethereum/go-ethereum/pull/13875.

There is not much reasoning on why; but since this commit is in that PR, I'd guess that it's because of the config file looking much more self-explanatory with

``` toml
SyncMode = "fast"
```

instead of

``` toml
Fast = true
```

This helps to avoid accidental overrides, the likes of

``` toml
SyncMode = "fast"
...
Light = true
```

which happens a lot with users cargo-culting when they "just want to blockchain real quick". (OT: Believe me, that happens [more than you'd want to know](https://github.com/ethereum/go-ethereum/issues/14461#issuecomment-351917303).)

This also applies to CLI args, not just config files.

-----

**TL;DR:** I'd recommend _not_ giving several ways to configure the same parameter. Having just one way reduces confusion.

**Comment by pipermerriam** [comment]
I'm still in favor of the pattern we have in place **for CLI options**.

- Assuming we have a flag that takes a concise number of well named string arguments (`--sync-mode fast/light/full`).
- It's very convenient to have quick shortcuts `--fast/--light/--full`

This doesn't need to apply to everything, but I think it's good UX (please tell me I'm wrong if you think so).

- Our CLI parser already handles the mutual exclusivity of these commands so that shouldn't be a problem.
- When we get around to doing config file stuff I don't plan on including those shortcuts so that shouldn't be a problem.

The main problem I see is when something like `--fast` can be easily conflated with meaning something else, but that is something we should be able to deal with on a case-by-case basis.

**Comment by cburgdorf** [comment]
>The main problem I see is when something like `--fast` can be easily conflated with meaning something else, but that is something we should be able to deal with on a case-by-case basis.

Another alternative would be to name the shorthands `--sync-fast`, `--sync-light` and `--sync-full`. Sits somewhere in between.

But this is just something that occurred to me. My brain has a very low resolution which basically eliminates all the subtle nuances of these UX issues. So, any solution is fine for :sweat_smile:

**Comment by carver** [comment]
> I'd recommend not giving several ways to configure the same parameter. Having just one way reduces the possibility confusion (and subsequent "bug" reports).

I tend to agree with this, but not so much that I would give a :-1: to a PR. So I guess I leave it up to the implementer.

> It's very convenient to have quick shortcuts

Other than as a trinity developer (when I'm repeatedly testing a wide range of options), I have always ended up with a long line of config parameters, written in a single-line shell file. For that scenario it seems nicer to have exactly one way to do things, and have them be explicit. So I'm not convinced that the quickness is a big enough benefit for end users, compared to the potential maintenance costs.

Also, it's much easier to add `--fast` later than it is to remove it later.

Anyhow, I don't think I would give a :-1: to a PR for including the shortcut. It's only a moderate opinion, and clearly smart people differ on it. So I guess it's up to the implementer, unless some broader consensus forms.

**Comment by pipermerriam** [comment]
I'm now convinced that the benefit isn't justified by the various downsides.  Lets remove the aliases.
